core_backend/validation/validate_retrieval.py

```diff
@@ -0,0 +1,275 @@
+    create_qdrant_collection,
+    get_qdrant_client,
+    get_search_results,
+)
+from core_backend.app.schemas import UserQueryBase, UserQuerySearchResult
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--validation_data_path", type=str, help="Path to validation data")
+parser.add_argument("--content_data_path", type=str, help="Path to validation data")
+parser.add_argument(
+    "--validation_data_question_col",
+    type=str,
+    help="Question column in validation data",
+)
+parser.add_argument(
+    "--validation_data_label_col",
+    type=str,
+    help="Content label column in validation data",
+)
+parser.add_argument(
+    "--content_data_label_col", type=str, help="Content label column in content data"
+)
+parser.add_argument(
+    "--content_data_text_col", type=str, help="Content text body column in content data"
+)
+parser.add_argument(
+    "--notification_topic", type=str, help="SNS topic to send notification to"
+)
+parser.add_argument(
+    "--n_similar",
+    "-n",
+    type=int,
+    help="Number of top similar content to retrieve",
+    default=10,
+)
+parser.add_argument(
+    "--aws_profile",
+    type=str,
+    help="AWS profile name",
+    default=None,
+)
+args = parser.parse_args()
+
+
+VALIDATION_COLLECTION_NAME = "validation_collection"
+
+
+async def aget_similar_content(
+    question: UserQueryBase,
+    qdrant_client: QdrantClient,
+    n_similar: int,
+    qdrant_collection_name: str,
+) -> Dict[int, UserQuerySearchResult]:
+    """
+    Get the most similar points in the vector db asynchronously
+    """
+    response = await aembedding(EMBEDDING_MODEL, question.query_text)
+    question_embedding = response.data[0]["embedding"]
+
+    return get_search_results(
+        question_embedding, qdrant_client, n_similar, qdrant_collection_name
+    )
+
+
+def get_rank(row: pd.Series) -> Union[int, None]:
+    """Get the rank of label in the retrieved content IDs"""
+    label = row[args.content_data_label_col]
+    ranked_list = row["retrieved_content_titles"]
+    try:
+        return ranked_list.index(label) + 1
+    except ValueError:
+        return None
+
+
+def get_retrieved_content_labels(
+    retrieval_results: Dict[int, UserQuerySearchResult]
+) -> List:
+    """Extract retrieved content IDs from results dict"""
+    i = 0
+    retrieved_content_labels = []
+    while i in retrieval_results:
+        label = retrieval_results[i].retrieved_title
+        retrieved_content_labels.append(label)
+        i += 1
+    return retrieved_content_labels
+
+
+def load_content_to_qdrant(content_dataframe: pd.DataFrame) -> QdrantClient:
+    """Load content to qdrant collection"""
+    client = get_qdrant_client()
+
+    if VALIDATION_COLLECTION_NAME not in {
+        collection.name for collection in client.get_collections().collections
+    }:
+        create_qdrant_collection(VALIDATION_COLLECTION_NAME, QDRANT_VECTOR_SIZE)
+
+    # get embeddings
+    embedding_results = embedding(
```

Comment:
Should we use `aembedding` again?

Reply:
I have added a new function called async_get_similar_content.. let me know what you think!